content/articles/tera-template-engine.md

```diff
@@ -0,0 +1,136 @@
+
+### Lexer/Parser
+This talk explains how the lexing in the [template package in the Go standard library](https://golang.org/pkg/text/template/) is implemented. I thought it was pretty cool and implemented something similar last summer.
+For those that don't want to watch the video, here's a quick summary.
+
+The lexer can be in a few states: inside a block, text, space, number, identifier, string for Tera currently and there are actions that represents what we do and result in a new state. In short, that means we have a state function that takes the lexer as an argument and returns a state function. While this is easy to do in Go, you cannot do reference the type while declaring it but it's ok to do so for a struct.
+```rust
+// working
+struct StateFn(Option<fn(&mut Lexer) -> StateFn>);
+// not working
+type StateFn = fn(&mut Lexer) -> Option<StateFn>;
+```
+Thanks for the help on IRC for that one.
+The lexer just runs the state function until we reach EOF which in our case is represented by returning `None` as a state function. Here's the whole run method of the lexer:
+
+```rust
+pub fn run(&mut self) {
+    loop {
+        let StateFn(state_fn) = self.state;
+        if state_fn.is_none() {
+            break;
+        }
+        self.state = state_fn.unwrap()(self);
+    }
+}
+```
+The actions read the next character and know what to do for each kind of character, eg. finding a number in a variable block will return the `lex_number` state function. The lexer ouput is a vector of tokens that ends with either a EOF or and error.
+
+You can read the whole lexer [on github](https://github.com/Keats/tera/blob/fddb8a0b82cba7374bd0552fed1cf831b8943395/src/lexer.rs), it is actually quite simple and readable.
+
+
+The parser is quite simple as well. Since we are either in text, in a variable block or in a tag block, we just handle those cases in a loop and use EOF to break. Each "state" knows how to parse itself so the main logic is actually only:
+
+```rust
+pub fn parse(&mut self) {
+    loop {
+        match self.peek().kind {
+            TokenType::TagStart => self.parse_tag_block(),
+            TokenType::VariableStart => self.parse_variable_block(),
+            TokenType::Text => self.parse_text(),
+            TokenType::Eof => break,
+            _ => unreachable!()
+        };
+    }
+}
+```
+
+The trickiest bit was handling precedence in blocks so that something like `{{1 / 2 + 3 * 2 + 42}}` would parse as expected. This is done by assigning precedence values to each kind of token and looking forward to see if something with higer precedence is coming. I would be surprised if there was not a bug in there.
+
+The output of the parser is a classic AST.
+
+### Context
+Context is the part where ease of use compared to dynamic languages really show. In python I can just pass a dict `{"user": user, "count": 1}` to Jinja2 or Django and be done with it. In Rust, it can't be that easy unfortunately.
```

Comment:
This is a bit misleading. Maybe "`Cotext` is where dynamic languages have the upper hand" ?